src/serena/util/file_system.py

```diff
@@ -138,21 +137,39 @@ def __init__(self, repo_root: str) -> None:
 
     def _load_gitignore_files(self) -> None:
         """Load all gitignore files from the repository."""
-        gitignore_files = self._find_gitignore_files()
-
-        for gitignore_file in gitignore_files:
-            spec = self._create_ignore_spec(gitignore_file)
+        for gitignore_path in self._iter_gitignore_files():
+            log.info("Processing gitignore file: %s", gitignore_path)
+            spec = self._create_ignore_spec(gitignore_path)
             if spec.patterns:  # Only add non-empty specs
                 self.ignore_specs.append(spec)
 
-    def _find_gitignore_files(self) -> list[str]:
+    @dataclass
+    class _QueueItem:
+        abs_path: str
+        rel_path: str | None = None
+
+    def _iter_gitignore_files(self, follow_symlinks: bool = False) -> Iterator[str]:
         """
-        Find all .gitignore files in the repository.
+        Iteratively discover .gitignore files in a top-down fashion, starting from the repository root.
+        Directory paths are skipped if they match any already loaded ignore patterns.
 
-        :return: List of absolute paths to .gitignore files
+        :return: an iterator yielding paths to .gitignore files (top-down)
         """
-        relative_paths = glob.glob("**/.gitignore", root_dir=self.repo_root, recursive=True)
-        return [os.path.join(self.repo_root, rel_path) for rel_path in relative_paths]
+        queue: list[GitignoreParser._QueueItem] = [self._QueueItem(abs_path=self.repo_root)]
+
+        def scan(abs_path: str, rel_path: str | None):
+            for entry in os.scandir(abs_path):
+                if entry.is_dir(follow_symlinks=follow_symlinks):
+                    dir_rel_path = rel_path + "/" + entry.name if rel_path else entry.name

```

Comment:
Is the truthy check on rel_path on purpose? If yes, a small comment on why empty string and None should be treated the same way would be nice. Though I don't see when empty string could ever be passed.

Also, shouldn't os.path.join be used here?